video/out/wayland_common.c

```diff
@@ -1022,17 +1022,14 @@ static void feedback_presented(void *data, struct wp_presentation_feedback *fbac
 {
     struct vo_wayland_state *wl = data;
 
-    // NULL is needed to prevent a dangling pointer since presentation_feedback
-    // is created in the frame_callback and not in any of the actual presentation
-    // events.
+    // NULL is needed to prevent a double free during uninit since wl->feedback will
+    // be the same as fback.
     if (fback) {
+        if (wl->feedback == fback)
+            wl->feedback = NULL;
```

Comment:
Is wl->feedback still useful at all?
Should the new logic not take care of all feedbacks?

Reply:
There's always one last feedback that's created and we won't get a discard or presented event. This essentially keeps track of that. The frame callback is the same way.

Reply:
Fair enough. Although it seems weird that this behavior would be intentional.

Reply:
I mean it's unavoidable. After every frame is drawn, we have to setup a new presentation feedback listener which will fire after the next draw occurs. If the user quits in the middle of this, well that next draw never actually occurs so you won't get a discarded or presented event from that. You have to just free it along with all the other `wl_proxy`'s.

Reply:
But that last listener is still registered to the surface, and destroying the surface should trigger a discard event.

Indeed adding a `wl_display_roundtrip(); wayland_dispatch_events();` after `wl_surface_destroy()` in `uninit()` seems to work.

I think the order has to be `roundtrip()` and then `dispatch()`, instead of the other way round as the PR does.
`roundtrip()` should make sure all events are delivered and `dispatch()` will handle them on mpv side.
Otherwise we may loose events (which I can reproduce)

Reply:
Then all of the `NULL` checks and special handling in the feedback listeners seem to be unnecessary, too.

Reply:
Maybe even better yet:
```c
if (roundtrip())
    dispatch()
```
This should make sure that we are never waiting in the dispatcher.

Reply:
This leaks for me if I move the surface to another desktop, go to that desktop, and then quit (sway). My PR leaks in that case as well though.

Reply:
Hm, it doesn't leak for me. With "desktop" do you mean virtual workspace or physical screen?

Could you push your version somewhere? Mine is here: https://github.com/t-8ch/mpv/tree/wayland-presentation-test

Reply:
Not really different than yours. Just this.

```
diff --git a/video/out/wayland_common.c b/video/out/wayland_common.c
index b1a3a6a2ff..bcac963285 100644
--- a/video/out/wayland_common.c
+++ b/video/out/wayland_common.c
@@ -1024,11 +1024,8 @@ static void feedback_presented(void *data, struct wp_presentation_feedback *fbac
 
     // NULL is needed to prevent a double free during uninit since wl->feedback will
     // be the same as fback.
-    if (fback) {
-        if (wl->feedback == fback)
-            wl->feedback = NULL;
+    if (fback)
         wp_presentation_feedback_destroy(fback);
-    }
 
     wl->refresh_interval = (int64_t)refresh_nsec / 1000;
 
@@ -1052,11 +1049,8 @@ static void feedback_discarded(void *data, struct wp_presentation_feedback *fbac
     struct vo_wayland_state *wl = data;
 
     // Same logic in feedback_presented applies here.
-    if (fback) {
-        if (wl->feedback == fback)
-            wl->feedback = NULL;
+    if (fback)
         wp_presentation_feedback_destroy(fback);
-    }
 }
 
 static const struct wp_presentation_feedback_listener feedback_listener = {
@@ -2153,9 +2147,6 @@ void vo_wayland_uninit(struct vo *vo)
     if (!wl)
         return;
 
-    wayland_dispatch_events(wl, 1, 10);
-    wl_display_roundtrip(wl->display);
-
     mp_input_put_key(wl->vo->input_ctx, MP_INPUT_RELEASE_ALL);
 
     if (wl->compositor)
@@ -2190,8 +2181,8 @@ void vo_wayland_uninit(struct vo *vo)
     if (wl->dnd_offer)
         wl_data_offer_destroy(wl->dnd_offer);
 
-    if (wl->feedback)
-        wp_presentation_feedback_destroy(wl->feedback);
+//    if (wl->feedback)
+//        wp_presentation_feedback_destroy(wl->feedback);
 
     if (wl->frame_callback)
         wl_callback_destroy(wl->frame_callback);
@@ -2251,6 +2242,9 @@ void vo_wayland_uninit(struct vo *vo)
     if (wl->video_subsurface)
         wl_subsurface_destroy(wl->video_subsurface);
 
+    if (wl_display_roundtrip(wl->display))
+        wayland_dispatch_events(wl, 1, 10);
+
     if (wl->wm_base)
         xdg_wm_base_destroy(wl->wm_base);
 
```

I meant virtual workspace my bad. If I change `10` to `100`, it doesn't leak.

Reply:
> If I change 10 to 100, it doesn't leak.

This feels a bit hacky. If I understand correctly the poll should return instantly anyways because we already forced all events through and they should be ready to be read.

It should even work with changing the 10 to 0, I think. It seems we are missing something.

Reply:
>This feels a bit hacky.

Yeah agreed. I think this is the right track though. If we can come up with something that isn't dependent on some timeout that always process the events, that should solve it.

Reply:
Putting in a `sleep(1)` before the dispatching makes the leak go away. So the discard events do all get sent.

Reply:
Also on mutter/gnome it works for me reliable even with timeout=0. Could you test that too?
Maybe a bug in wlroots?

Reply:
Indeed, it seems to always work in mutter.

Reply:
Some unfortunate news, this method doesn't seem to work in weston at all. Sway will eventually deliver the event if you wait long enough (in this case, the second-to-last feedback creation) so in theory a better block would solve it. In weston this doesn't appear to happen at all. It never sends a discard.

Reply:
But that would mean on weston presentation-time will leak unbounded amounts of memory anyways.
This seems like a fairly serious bug. Cleaning it up properly at `uninit()` is a very weak bandaid.

Reply:
Oh sorry, I phrased that badly. I meant the last discard isn't cleaned (AKA it will always leak the same amount of memory anyway). However, I think this one is actually a weston bug. The code looks like it attempts to send a discard when the surface is destroyed but we never get it for some reason. With sway/wlroots, I'm not sure exactly what's going on. It gives us the discard immediately but for some unspecified period of time later, it gives us the second to last presented event (which occurs after the discarded one). Ironically, mutter is the one that works as I would expect. Go figure.

Reply:
3c62bee works nicely for me on sway.

Reply:
Not for me. Unless I bump up the timeout, but I'm trying to avoid that (no luck yet).

Reply:
FYI the behavior of weston is definitively a bug. It leaks the surface in the fade animation somewhere.

Reply:
Nice. I'm also pretty sure that sway/wlroots is bugged too here. When destroying the surface (aka quitting mpv), you get logs like this (present/discard added by me).

```
00:00:07.771 [DEBUG] [sway/desktop/idle_inhibit_v1.c:22] Sway idle inhibitor destroyed
00:00:07.771 [DEBUG] [wlr] [types/wlr_idle.c:183] Enabling idle timers for all seats
00:00:07.771 [DEBUG] [sway/desktop/idle_inhibit_v1.c:22] Sway idle inhibitor destroyed
00:00:07.771 [DEBUG] [sway/tree/arrange.c:263] Usable area for ws: 1920x1059@0,0
00:00:07.771 [DEBUG] [sway/tree/arrange.c:293] Arranging workspace '1' at 0.000000, 0.000000
00:00:07.771 [DEBUG] [sway/tree/arrange.c:77] Arranging 0x7fffa441fa50 horizontally
00:00:07.771 [DEBUG] [sway/desktop/transaction.c:398] Transaction 0x55cd5a8564e0 committing with 5 instructions
00:00:07.771 [DEBUG] [sway/desktop/transaction.c:294] Applying transaction 0x55cd5a8564e0
00:00:07.771 [INFO] [wlr] [types/wlr_presentation_time.c:244] Send discard event.
00:00:07.779 [INFO] [wlr] [types/wlr_presentation_time.c:300] Send present event.
```
That final present event (which corresponds to the second-to-last feedback mpv creates) gets sent *after* sway commits its atomic transaction for surface destruction. That's why adding a slight timeout works. When you screw around with the surface, the time it takes to send the final present event can increase. When a surface is destroyed, it should send any presented events at the same time as discarded imo.

tl;dr: I think this patch is right now but there's a couple of compositor bugs (I'll try to fix wlroots).

Reply:
Do you already have a simple reproduction case?